src/HessianNum.jl

```diff
@@ -0,0 +1,179 @@
+end
+
+function promote_rule{N,T1,C1,T2,C2}(::Type{HessianNum{N,T1,C1}}, ::Type{HessianNum{N,T2,C2}})
+    R = promote_type(T1, T2)
+    return HessianNum{N,R,switch_eltype(C1, R)}
+end
+
+#######################
+# Math on HessianNums #
+#######################
+
+## Bivariate functions on HessianNums ##
+##------------------------------------##
+
+const h_bivar_funcs = Tuple{Symbol, Expr}[
+    (:*, :(hess(a,k)*value(b)+grad(a,i)*grad(b,j)+grad(a,j)*grad(b,i)+value(a)*hess(b,k))),
+    (:/, :(((2*value(a)*grad(b,j)*grad(b,i)+(value(b)^2)*hess(a,k))-(value(b)*(grad(a,i)*grad(b,j)
+           +grad(a,j)*grad(b,i)+value(a)*hess(b,k))))/(value(b)^3))),
+    (:^, :((value(a)^(value(b)-2))*((value(b)^2)*grad(a,i)*grad(a,j)+value(b)*(grad(a,j)*(-grad(a,i)
+           +value(a)*log(value(a))*grad(b,i))+value(a)*(log(value(a))*grad(a,i)*grad(b,j)
+           +hess(a,k)))+value(a)*(grad(a,j)*grad(b,i)+grad(b,j)*(grad(a,i)+value(a)*log(value(a))
+           *log(value(a))*grad(b,i))+value(a)*log(value(a))*hess(b,k)))))
+]
+
+for (fsym, term) in h_bivar_funcs
+    loadfsym = symbol(string("loadhess_", fsym, "!"))
+    @eval begin
+        function $(loadfsym){N}(a::HessianNum{N}, b::HessianNum{N}, output)
+            k = 1
+            for i in 1:N
+                for j in 1:i
+                    output[k] = $(term)
+                    k += 1
+                end
+            end
+            return output
+        end
+
+        function $(fsym){N,A,B}(a::HessianNum{N,A}, b::HessianNum{N,B})
+            new_hess = Array(promote_type(A, B), halfhesslen(N))
+            return HessianNum($(fsym)(grad(a), grad(b)), $(loadfsym)(a, b, new_hess))
+        end
+
+    end
+end
+
++{N}(a::HessianNum{N}, b::HessianNum{N}) = HessianNum(grad(a) + grad(b), hess(a) + hess(b))
+-{N}(a::HessianNum{N}, b::HessianNum{N}) = HessianNum(grad(a) - grad(b), hess(a) - hess(b))
+
+for T in (:Bool, :Real)
+    @eval begin
+        *(h::HessianNum, x::$(T)) = HessianNum(grad(h) * x, hess(h) * x)
+        *(x::$(T), h::HessianNum) = HessianNum(x * grad(h), x * hess(h))
+    end
+end
+
+/(h::HessianNum, x::Real) = HessianNum(grad(h) / x, hess(h) / x)
+#/(x::Real, h::HessianNum) = ?
+
+for T in (:Rational, :Integer, :Real)
+    @eval begin
+        function ^{N}(h::HessianNum{N}, p::$(T))
+            new_hess = Array(promote_type(eltype(h), typeof(p)), halfhesslen(N))
+            k = 1
+            for i in 1:N
+                for j in 1:i
+                    new_hess[k] = p*value(h)^(p-2)*((p-1)*grad(h,i)*grad(h,j)+value(h)*hess(h,k))
+                    k += 1
+                end
+            end
+            return HessianNum(grad(h)^p, new_hess)
+        end
+    end
+end
+
+## Univariate functions on HessianNums ##
+##-------------------------------------##
+
+-(h::HessianNum) = HessianNum(-grad(h), -hess(h))
+
+const h_univar_funcs = Tuple{Symbol, Expr}[
+    (:sqrt, :((-grad(h,i)*grad(h,j)+2*value(h)*hess(h,i)) / (4*(value(h)^(1.5))))),
```

Comment:
Could you document the format for this expression table? It's not really clear to me what's going on